From db849df4c05690a2016fb2eb47de9badd8aad84b Mon Sep 17 00:00:00 2001
From: Andreas Kurth <adk@lowrisc.org>
Date: Fri, 21 Mar 2025 19:35:18 +0000
Subject: [PATCH] [prim] Add single-cycle variant of `prim_ram_1p_scr`

Ibex (data and tags of instruction cache) and OTBN (data and instruction
memory) use `prim_ram_1p_scr` in a particular way:  They ignore the
`gnt_o` (and, in the case of Ibex, also the `rvalid_o`) signal, hence
assuming that `prim_ram_1p_scr` is always able to handle requests and
will provide a response in the next cycle.  This is true for
the current implementation of `prim_ram_1p_scr`, but it means `gnt_o`
and `rvalid_o` cannot be used for flow control, which is problematic for
upcoming changes to `prim_ram_1p_scr`.

This commit thus creates a variant of `prim_ram_1p_scr` with a `_1cyc`
suffix.  This variant is always ready for requests and always responds
in the next cycle.

As there's a circular dependency on this change with the Ibex repo,
which gets vendored into this repo, this commit applies a patch to Ibex.
Once this commit has been merged and the Ibex repo has been updated,
that patch can be removed.

Signed-off-by: Andreas Kurth <adk@lowrisc.org>
---
 ibex_top.sv | 10 ++--------
 1 file changed, 2 insertions(+), 8 deletions(-)

diff --git a/ibex_top.sv b/ibex_top.sv
index 7ed24b4a59..a6be38fc60 100644
--- a/ibex_top.sv
+++ b/ibex_top.sv
@@ -591,7 +591,7 @@ module ibex_top import ibex_pkg::*; #(
 
         // SEC_CM: ICACHE.MEM.SCRAMBLE
         // Tag RAM instantiation
-        prim_ram_1p_scr #(
+        prim_ram_1p_scr_1cyc #(
           .Width              (TagSizeECC),
           .Depth              (IC_NUM_LINES),
           .DataBitsPerMask    (TagSizeECC),
@@ -602,13 +602,11 @@ module ibex_top import ibex_pkg::*; #(
           .clk_i,
           .rst_ni,
 
-          .key_valid_i      (scramble_key_valid_q),
           .key_i            (scramble_key_q),
           .nonce_i          (scramble_nonce_q),
 
           .req_i            (ic_tag_req[way]),
 
-          .gnt_o            (),
           .write_i          (ic_tag_write),
           .addr_i           (ic_tag_addr),
           .wdata_i          (ic_tag_wdata),
@@ -616,7 +614,6 @@ module ibex_top import ibex_pkg::*; #(
           .intg_error_i     (1'b0),
 
           .rdata_o          (ic_tag_rdata[way]),
-          .rvalid_o         (),
           .raddr_o          (),
           .rerror_o         (),
           .cfg_i            (ram_cfg_icache_tag_i),
@@ -628,7 +625,7 @@ module ibex_top import ibex_pkg::*; #(
         );
 
         // Data RAM instantiation
-        prim_ram_1p_scr #(
+        prim_ram_1p_scr_1cyc #(
           .Width              (LineSizeECC),
           .Depth              (IC_NUM_LINES),
           .DataBitsPerMask    (LineSizeECC),
@@ -640,13 +637,11 @@ module ibex_top import ibex_pkg::*; #(
           .clk_i,
           .rst_ni,
 
-          .key_valid_i      (scramble_key_valid_q),
           .key_i            (scramble_key_q),
           .nonce_i          (scramble_nonce_q),
 
           .req_i            (ic_data_req[way]),
 
-          .gnt_o            (),
           .write_i          (ic_data_write),
           .addr_i           (ic_data_addr),
           .wdata_i          (ic_data_wdata),
@@ -654,7 +649,6 @@ module ibex_top import ibex_pkg::*; #(
           .intg_error_i     (1'b0),
 
           .rdata_o          (ic_data_rdata[way]),
-          .rvalid_o         (),
           .raddr_o          (),
           .rerror_o         (),
           .cfg_i            (ram_cfg_icache_data_i),
-- 
2.34.1

